piccolo_cursor_pagination/pagination.py

```diff
@@ -0,0 +1,153 @@
+import base64
+import typing as t
+from dataclasses import dataclass
+
+from fastapi.requests import Request
+from piccolo.table import Table
+
+
+@dataclass
+class CursorPagination:
+    """
+    Used to cursor pagination.
+
+    :param cursor:
+        Value of cursor as a string.
+    :param page_size:
+        Value of page_size. Default to ``15``.
+    :param order_by:
+        Value of order_by descending order
+        Default to ``-id``.
+    """
+
+    cursor: str
+    page_size: int = 15
+    order_by: str = "-id"
+
+    def get_cursor_rows(self, table: t.Type[Table], request: Request):
```

Comment:
I would consider moving these values around a bit. So instead it's:

```python
@dataclass
class CursorPagination:
    table: t.Type[Table]
    page_size: int = 15
    order_by: str = "-id"

    def get_cursor_rows(self, cursor: str, request: Request):
        ...
```

This means you could create `CursorPagination` as a global, and reuse it for each request.

```python
PAGINATION = CursorPagination(table=Movie)

def my_endpoint(request):
    data = PAGINATION.get_cursor_rows(cursor, request)
    ...
```

Also, you might be able to decouple it from Starlette / FastAPI etc if rather than passing in `Request`, you pass in `previous_cursor` instead. This means it could work with BlackSheep too.

Reply:
You could even have it as a function instead of a class:

```python
def get_cursor_rows(cursor: str, request: Request, table, page_size, order_by):
   ...
```

I'll stop interfering ... just think about what needs to be class variables vs function arguments.

Reply:
Thanks for your feedback, I appreciate it.
> I would consider moving these values around a bit. So instead it's:
> 
> ```python
> @dataclass
> class CursorPagination:
>     table: t.Type[Table]
>     page_size: int = 15
>     order_by: str = "-id"
> 
>     def get_cursor_rows(self, cursor: str, request: Request):
>         ...
> ```
> 
> This means you could create `CursorPagination` as a global, and reuse it for each request.
> 
> ```python
> PAGINATION = CursorPagination(table=Movie)
> 
> def my_endpoint(request):
>     data = PAGINATION.get_cursor_rows(cursor, request)
>     ...
> ```
Unfortunately, I failed to make this work.
> Also, you might be able to decouple it from Starlette / FastAPI etc if rather than passing in Request, you pass in  previous_cursor instead. This means it could work with BlackSheep too.

This would be nice but we need `Request` to get `request.query_params["__cursor"]`and `request.query_params["__previous"]`

> You could even have it as a function instead of a class:
> 
> ```python
> def get_cursor_rows(cursor, request, table, page_size, order_by):
>    ...
> ```
> 
> I'll stop interfering ... just think about what needs to be class variables vs function arguments.

You are right, but I think this structure is good.
```python
paginator = CursorPagination(cursor=cursor) # instantiation
rows_result, headers_result = paginator.get_cursor_rows(Task, request) # unpacking values from method
```
I think this is solid and easy to understand.
Thanks again.